bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/discovery/UsbGatewayDiscoveryService.java

```diff
@@ -0,0 +1,246 @@
+    private @Nullable USBGateway zbGateway;
+
+    private String currentScannedPortName = "";
+
+    /**
+     * Keeps a boolean during time discovery process in progress.
+     */
+    private boolean scanning;
+
+    /**
+     * Constructs a new UsbGatewayDiscoveryService with the specified ZigBee USB Bridge ThingTypeUID
+     */
+    public UsbGatewayDiscoveryService() {
+        super(new HashSet<>(Arrays.asList(OpenWebNetBindingConstants.THING_TYPE_ZB_GATEWAY)), DISCOVERY_TIMEOUT_SECONDS,
+                false);
+    }
+
+    /**
+     * Starts a new discovery scan. All available Serial Ports are scanned.
+     */
+    @Override
+    protected void startScan() {
+        logger.debug("Started OpenWebNet ZigBee USB Gateway discovery scan");
+        removeOlderResults(getTimestampOfLastScan());
+        scanning = true;
+        Stream<SerialPortIdentifier> portEnum = serialPortManager.getIdentifiers();
+        // Check each available serial port
+        portEnum.forEach(portIdentifier -> {
+            if (scanning) {
+                currentScannedPortName = portIdentifier.getName();
+                logger.debug("[{}] == checking serial port", currentScannedPortName);
+                if (portIdentifier.isCurrentlyOwned()) {
+                    logger.debug("[{}] serial port is owned by: {}", currentScannedPortName,
+                            portIdentifier.getCurrentOwner());
+                    if ("NRSerialPort".equals(portIdentifier.getCurrentOwner())) {
+                        logger.debug(
+                                "[{}] serial port is owned by this binding (or the openwebnet4j/nrjavaserial libs). If no ZigBee USB Gateway is already configured on this port it might indicate the port is locked by an older instance of this binding. Restart the system to unlock the port.",
+                                currentScannedPortName);
+                    }
+                } else {
+                    logger.debug("[{}] trying to connect to a ZigBee USB Gateway...", currentScannedPortName);
+                    USBGateway gw = new USBGateway(currentScannedPortName);
+                    zbGateway = gw;
+                    gw.subscribe(this);
+                    portCheckLatch = new CountDownLatch(1);
+                    connectTimeout = scheduler.schedule(() -> {
+                        logger.debug("[{}] timeout expired", currentScannedPortName);
+                        endSerialPortScan();
+                        portCheckLatch.countDown();
+                    }, PORT_CHECK_TIMEOUT_MSEC, TimeUnit.MILLISECONDS);
+                    try {
+                        gw.connect();
+                        try { // wait for onConnected/onConnectionError or connectTimeout
+                            portCheckLatch.await();
+                        } catch (InterruptedException ie) {
+                            logger.warn("[{}] interrupted: {}", currentScannedPortName, ie.getMessage());
+                        }
```

Comment:
The program flow should finish as quickly as possible when an InterruptedException is raised. You could make the foreach a conventional `for` and catch the exception outside of the loop.